HttpSemanticConventions - new metric in AspNetCore Instrumentation. (#4802)
open-telemetry/opentelemetry-dotnet: src/OpenTelemetry.Instrumentation.AspNetCore/README.md

#### List of metrics produced

The instrumentation is implemented based on
[metrics semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/4bbb8c907402caa90bc077214e8a2c78807c1ab9/docs/http/http-metrics.md).

> Have you opt-ed into the new Http Semantic Conventions?

Reviewer: nit: Mention the environment variable name `OTEL_SEMCONV_STABILITY_OPT_IN` here for clarity.

Author: done.

* If yes, the instrumentation supports the following metric.

    | Name  | Instrument Type | Unit | Description |
    |-------|-----------------|------|-------------|

>     | `http.server.request.duration` | Histogram | `s` | Measures the duration of inbound HTTP requests. |

Reviewer 2: it'd be good to callout here that SDK uses a special treatment for this metric by using a suitable buckets.
https://github.com/open-telemetry/opentelemetry-dotnet/pull/4820/files

Author: How much information are you looking for?
Is it sufficient to say "http.server.request.duration uses unique buckets suitable for seconds" ?

I could include a link to the spec which shows the buckets:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpserverrequestduration

Reviewer 2: something along the lines of
"This metric is emitted in "s" as unit, as per the semantic convention guide. The convention also recommends to provide a custom buckets (link), but that capability is not yet available in OTel .NET. A workaround is already included in the OTel SDK (link), which specially treats this metric, and uses the recommended buckets"

^ feel free to edit/modify!

Author: updated. please re-review

Reviewer 2: I'd highlight/make it clear that the workaround is "OpenTelemetry SDK specially treats this metric and uses a diff. default for histogram buckets, matching the recommended buckets" /similar wording. This will ensure most folks do not have any need to go and read the linked PR.

Reviewer 2: the linked PR is very light on details. That must be fixed separately. Specifically, 
1. it should list the exact name of the metrics for which SDK uses a diff. default.
2. make it clear that the special treatment is only for the default case. If user overrides the buckets, then that'll always be respected. 

(Not part of this PR, but this is essential to get the whole transition smoother for end users)

#### List of metrics produced

The instrumentation is implemented based on
[metrics semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/4bbb8c907402caa90bc077214e8a2c78807c1ab9/docs/http/http-metrics.md).

Have you opt-ed into the new Http Semantic Conventions using `OTEL_SEMCONV_STABILITY_OPT_IN`?

* If yes, the instrumentation supports the following metric.

    | Name  | Instrument Type | Unit | Description |
    |-------|-----------------|------|-------------|
    | `http.server.request.duration` | Histogram | `s` | Measures the duration of inbound HTTP requests. |

    This metric is emitted in "seconds" as per the semantic convention. While
    the convention [recommends using custom histogram buckets](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpserverrequestduration)

>     , this feature is not yet available in OTel .NET. However, the OTel SDK

Reviewer: ```suggestion
    , this feature is not yet available via .NET metrics API.  A workaround has been included in OTel SDK starting version `1.6.0` which applies recommended buckets by default for this metric.
```

Reviewer: Also, this needs to be included in changelog: No action is needed from users targeting latest version of instrumentation library as that will bring in 1.6.0 SDK version by default.

#### List of metrics produced

The instrumentation is implemented based on
[metrics semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/4bbb8c907402caa90bc077214e8a2c78807c1ab9/docs/http/http-metrics.md).

> Have you opt-ed into the new Http Semantic Conventions using `OTEL_SEMCONV_STABILITY_OPT_IN`?

Reviewer 3: ```suggestion
Have you opted in to the new Http Semantic Conventions using `OTEL_SEMCONV_STABILITY_OPT_IN`?
```

> * If yes, the instrumentation supports the following metric.

Reviewer 2: this is not clear. The env variable can take none,http,http/dup as values, This doc should also be structured in that way. i.e what if the env variable is set to "none", (also the default)
what is the env varaible is set to "http" 
what is the env varaible is set to "http/dup" - both metric are emitted

    | Name  | Instrument Type | Unit | Description |
    |-------|-----------------|------|-------------|
    | `http.server.request.duration` | Histogram | `s` | Measures the duration of inbound HTTP requests. |

>     This metric is emitted in "seconds" as per the semantic convention. While

Reviewer: ```suggestion
    This metric is emitted in `seconds` as per the semantic convention. While
```

>     This metric is emitted in `seconds` as per the semantic convention. While

Reviewer 3: ```suggestion
    This metric value is emitted in `seconds` as per the semantic convention. While
```

>     the convention [recommends using custom histogram buckets](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpserverrequestduration)

Reviewer 3: Use the commit tag?

Author: I'm conflicted here. 
I would rather link to an explicit tag. There changes we're implementing now are post v1.21.0, but this repo hasn't made a new release yet.

I can change links in both the Readme and Changelog to a commit as you've suggested.
When the next release happens (presumably v1.22.0) I think we should update these links.

#### List of metrics produced

> A different metric is emitted depending if a user opts-in to the new Http

Reviewer 3: ```suggestion
A different metric is emitted depending on whether a user opts-in to the new Http
```
